File: android/src/main/kotlin/com/andresmontano/native_haptics_and_audio/NativeHapticsAndAudioPlugin.kt
package com.andresmontano.native_haptics_and_audio

import android.content.Context
import android.media.AudioAttributes
import android.media.SoundPool
import android.os.Build
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager
import io.flutter.embedding.engine.plugins.FlutterPlugin
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * Android implementation of the [HapticsAndAudioApi] Pigeon interface.
 *
 * Uses [SoundPool] for zero-latency PCM playback and the [Vibrator]
 * service for native haptic feedback patterns.
 */
class NativeHapticsAndAudioPlugin :
    FlutterPlugin,
    HapticsAndAudioApi {

    private lateinit var context: Context
    private var soundPool: SoundPool? = null
    private val soundIds = mutableMapOf<PosSound, Int>()

    // ──────────────────────────────────────────────
    // FlutterPlugin lifecycle
    // ──────────────────────────────────────────────

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        HapticsAndAudioApi.setUp(binding.binaryMessenger, this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        HapticsAndAudioApi.setUp(binding.binaryMessenger, null)
        releaseSoundPool()
    }

    // ──────────────────────────────────────────────
    // HapticsAndAudioApi – Pigeon interface
    // ──────────────────────────────────────────────

    /**
     * Pre-loads all `.wav` assets from `res/raw` into the [SoundPool].
     *
     * Disk I/O runs on [Dispatchers.IO]; the Pigeon [callback] is invoked
     * only after every asset has been fully loaded into native RAM.
     */
    override fun initialize(callback: (Result<Unit>) -> Unit) {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val audioAttributes = AudioAttributes.Builder()
                    .setUsage(AudioAttributes.USAGE_ASSISTANCE_SONIFICATION)
                    .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
                    .build()

                val pool = SoundPool.Builder()
                    .setMaxStreams(4)
                    .setAudioAttributes(audioAttributes)
                    .build()

                soundPool = pool

                // Map each enum value to its corresponding raw resource.
                val resourceMap = mapOf(
                    PosSound.SCANNER_BEEP to R.raw.store_scanner_beep,
                    PosSound.WARNING_BEEP to R.raw.beep_warning,
                    PosSound.DOUBLE_WARNING_BEEP to R.raw.double_beep_warning,
                    PosSound.KACHING to R.raw.kaching,
                )

                for ((sound, resId) in resourceMap) {
                    soundIds[sound] = pool.load(context, resId, 1)
                }

                // Give SoundPool a moment to finish internal decoding.
                // SoundPool.load() is async internally; a brief yield
                // ensures buffers are ready for instant playback.
                kotlinx.coroutines.delay(150)

                withContext(Dispatchers.Main) {
                    callback(Result.success(Unit))
                }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    callback(Result.failure(e))
                }
            }
        }
    }

    /** Instantly plays a pre-loaded audio asset at full volume. */
    override fun playSound(sound: PosSound) {
        val id = soundIds[sound]
            ?: throw FlutterError("SOUND_NOT_LOADED", "Sound $sound has not been loaded.")
        soundPool?.play(id, 1.0f, 1.0f, 1, 0, 1.0f)
    }

    /**
     * Triggers a native haptic vibration pattern.
     *
     * - [PosHaptic.SUCCESS] → short tick  (20 ms)
     * - [PosHaptic.WARNING] → medium buzz (100 ms)
     * - [PosHaptic.ERROR]   → heavy double-buzz (0, 80, 100, 80 ms pattern)
     */
    override fun playHaptic(haptic: PosHaptic) {
        val vibrator = getVibrator()

        when (haptic) {
            PosHaptic.SUCCESS -> {
                vibrator.vibrate(
                    VibrationEffect.createOneShot(65, 255)
                )
            }
            PosHaptic.WARNING -> {
                vibrator.vibrate(
                    VibrationEffect.createOneShot(100, VibrationEffect.DEFAULT_AMPLITUDE)
                )
            }
            PosHaptic.ERROR -> {
                vibrator.vibrate(
                    VibrationEffect.createWaveform(
                        longArrayOf(0, 80, 100, 80),
                        intArrayOf(0, 255, 0, 255),
                        -1
                    )
                )
            }
        }
    }

    /** Releases all [SoundPool] resources and clears the loaded sound map. */
    override fun release() {
        releaseSoundPool()
    }

    // ──────────────────────────────────────────────
    // Private helpers
    // ──────────────────────────────────────────────

    private fun releaseSoundPool() {
        soundPool?.release()
        soundPool = null
        soundIds.clear()
    }

    @Suppress("DEPRECATION")
    private fun getVibrator(): Vibrator {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            val vibratorManager =
                context.getSystemService(Context.VIBRATOR_MANAGER_SERVICE) as VibratorManager
            vibratorManager.defaultVibrator
        } else {
            context.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
        }
    }
}

File: ios/Classes/NativeHapticsAndAudioPlugin.swift
import Flutter
import UIKit
import AudioToolbox
import AVFoundation

/// iOS implementation of the `HapticsAndAudioApi` Pigeon protocol.
///
/// Uses `AudioServicesCreateSystemSoundID` for zero-latency PCM playback
/// and `UIImpactFeedbackGenerator` / `UINotificationFeedbackGenerator` for native haptic feedback.
public class NativeHapticsAndAudioPlugin: NSObject, FlutterPlugin, HapticsAndAudioApi {

    // MARK: - Properties

    private var soundIds: [PosSound: SystemSoundID] = [:]

    /// Maps each `PosSound` enum value to its corresponding `.wav` filename.
    private static let soundFileNames: [PosSound: String] = [
        .scannerBeep: "store_scanner_beep",
        .warningBeep: "beep_warning",
        .doubleWarningBeep: "double_beep_warning",
        .kaching: "ka-ching",
    ]

    // MARK: - FlutterPlugin registration

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = NativeHapticsAndAudioPlugin()
        HapticsAndAudioApiSetup.setUp(
            binaryMessenger: registrar.messenger(),
            api: instance
        )
    }

    // MARK: - HapticsAndAudioApi

    /// Pre-loads all `.wav` assets into RAM via `AudioServicesCreateSystemSoundID`.
    ///
    /// The resource bundle lookup follows a two-tier strategy:
    /// 1. Look in the CocoaPods-generated resource bundle (`native_haptics_and_audio`).
    /// 2. Fall back to `Bundle.main` if the assets were flattened by the host build system.
    func initialize(completion: @escaping (Result<Void, Error>) -> Void) {
        do {
            // --- Resolve the resource bundle ---
            let resourceBundle = resolveResourceBundle()

            // --- Load each .wav into an AudioServices SystemSoundID ---
            for (sound, fileName) in Self.soundFileNames {
                guard let url = resourceBundle.url(
                    forResource: fileName,
                    withExtension: "wav"
                ) else {
                    throw PigeonError(
                        code: "ASSET_NOT_FOUND",
                        message: "Could not locate \(fileName).wav in the plugin bundle.",
                        details: nil
                    )
                }

                var systemSoundId: SystemSoundID = 0
                let status = AudioServicesCreateSystemSoundID(url as CFURL, &systemSoundId)
                guard status == kAudioServicesNoError else {
                    throw PigeonError(
                        code: "AUDIO_LOAD_FAILED",
                        message: "AudioServicesCreateSystemSoundID failed for \(fileName).wav with status \(status).",
                        details: nil
                    )
                }

                soundIds[sound] = systemSoundId
            }

            // Haptic generators are lightweight and instantiated on demand in playHaptic().

            completion(.success(()))
        } catch {
            completion(.failure(error))
        }
    }

    /// Instantly plays the pre-loaded system sound for the given [sound].
    func playSound(sound: PosSound) throws {
        guard let systemSoundId = soundIds[sound] else {
            throw PigeonError(
                code: "SOUND_NOT_LOADED",
                message: "Sound \(sound) has not been loaded. Call initialize() first.",
                details: nil
            )
        }
        AudioServicesPlaySystemSound(systemSoundId)
    }

    /// Triggers a native haptic pattern matching the given [haptic] type.
    ///
    /// Uses `UIImpactFeedbackGenerator(.rigid)` for a crisp `success` tick,
    /// and `UINotificationFeedbackGenerator` for heavier `warning` / `error` pulses.
    func playHaptic(haptic: PosHaptic) throws {
        switch haptic {
        case .success:
            // Single, ultra-crisp mechanical tick.
            let impactGen = UIImpactFeedbackGenerator(style: .rigid)
            impactGen.prepare()
            impactGen.impactOccurred()

        case .warning:
            // Heavier notification pulse.
            let notifGen = UINotificationFeedbackGenerator()
            notifGen.prepare()
            notifGen.notificationOccurred(.warning)

        case .error:
            // Sustained, heavy double-pulse.
            let notifGen = UINotificationFeedbackGenerator()
            notifGen.prepare()
            notifGen.notificationOccurred(.error)
        }
    }

    /// Disposes all loaded `SystemSoundID` handles.
    func release() throws {
        for (_, systemSoundId) in soundIds {
            AudioServicesDisposeSystemSoundID(systemSoundId)
        }
        soundIds.removeAll()
    }

    // MARK: - Private helpers

    /// Resolves the resource bundle for this plugin.
    ///
    /// - Under **SwiftPM**: uses `Bundle.module`, which is synthesized automatically
    ///   by the Swift build system for the package's declared resources.
    /// - Under **CocoaPods**: uses the `resource_bundles` nested bundle strategy,
    ///   with a `Bundle.main` fallback for hosts that flatten assets.
    private func resolveResourceBundle() -> Bundle {
        #if SWIFT_PACKAGE
        // SwiftPM synthesizes Bundle.module for the package target's resources.
        return Bundle.module
        #else
        // The CocoaPods `resource_bundles` directive creates a nested bundle
        // named after the key in the podspec (`native_haptics_and_audio`).
        let frameworkBundle = Bundle(for: NativeHapticsAndAudioPlugin.self)
        if let resourceBundleURL = frameworkBundle.url(
            forResource: "native_haptics_and_audio",
            withExtension: "bundle"
        ), let resourceBundle = Bundle(url: resourceBundleURL) {
            return resourceBundle
        }

        // Fallback: assets may have been flattened into the main app bundle.
        return Bundle.main
        #endif
    }
}
